### PR TITLE
Upgrade TempFile implementation, independ from /proc/self/fd/x

### PR DESCRIPTION
**Pull request by qiuxiafei**

### CODE REVIEW

File: pytorch_blade/src/common_utils/tempfs.cpp [hunk 1]
```diff
@@ -11,57 +11,93 @@
...
+    ::unlink(fname_.c_str());
+  }
+  if (fd_ > 0) {
+    ::close(fd_);
+  }
 }
 
 bool TempFile::WriteBytesToFile(const std::string& bytes) {
-  auto sz = std::fwrite(bytes.data(), sizeof(char), bytes.size(), tmp_file_);
+  auto sz = ::write(fd_, bytes.data(), bytes.length());
```
Comment (fortianyou) [inline pytorch_blade/src/common_utils/tempfs.cpp:78]: I want to make sure there is no need to flush buffer here?

File: pytorch_blade/src/common_utils/tempfs.cpp [hunk 2]
```diff
@@ -11,57 +11,93 @@
...
+  // http://pubs.opengroup.org/onlinepubs/009695399/functions/mkstemp.html
+  static const std::string kRandomPattern = "XXXXXX";
+
+  // We see if any of these environment variables is set and use their value, or
+  // else default the temporary directory to `/tmp`.
+  static const char* env_variables[] = {"TMPDIR", "TMP", "TEMP", "TEMPDIR"};
+
+  std::string tmp_directory = "/tmp";
+  for (const char* variable : env_variables) {
+    if (const char* path = getenv(variable)) {
```
[the remark this answers is not in the document]
Comment (qiuxiafei) [reply pytorch_blade/src/common_utils/tempfs.cpp]: fixed

File: pytorch_blade/src/common_utils/tempfs.cpp [hunk 3]
```diff
@@ -11,57 +11,93 @@
...
+      tmp_directory = path;
+      break;
+    }
+  }
+
+  std::vector<char> filename;
```
[the remark this answers is not in the document]
Comment (qiuxiafei) [reply pytorch_blade/src/common_utils/tempfs.cpp:46]: `ostringstream` creates a `std::string`, but the `mkstemp` function required a modifiable `char *` to set the temp file name, while `std::string` have only `const char *` members (before c++17).

[the remark this answers is not in the document]
Comment (qiuxiafei) [reply pytorch_blade/src/common_utils/tempfs.cpp:46]: `ostringstream` creates a `std::string`, but the `mkstemp` function required a modifiable `char *` to set the temp file name, while `std::string` have only `const char *` members (before c++17).

File: pytorch_blade/src/common_utils/tempfs.cpp (hunk 1 above)
Comment (qiuxiafei) [reply pytorch_blade/src/common_utils/tempfs.cpp:80]: `write` is a unbuffered system call while `fwrite` is a buffered C library call.

**Review by fortianyou — APPROVED** [review]
LGTM

**Pull request merged**